### PR TITLE
Allow to select an existing security to replace a newly imported one

### PR DESCRIPTION
**Pull request by buchen**
When reviewing the newly created items, the user can choose to replace a new security with an existing security. All transactions are then imported and added to the existing security.

* Extended ExtractedEntry with security dependency (= if an extracted entry depends on another extracted security) and security override (= if the user chose to replace the imported security with an existing security)
* Refactored the import logic into ImportController to allow sharing between PDF and CSV imports

https://github.com/portfolio-performance/portfolio/assets/587976/70d37082-d2f7-4d27-932a-2601c63f0a11

### CODE REVIEW

File: name.abuchen.portfolio.ui/src/name/abuchen/portfolio/ui/wizards/datatransfer/ImportController.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+        boolean isDirty = false;
+        for (int index = 0; index < pages.size(); index++)
+        {
+            ReviewExtractedItemsPage page = pages.get(index);
+            page.afterPage();
+
+            var isPageDirty = importPage(page, newSecurities);
+            isDirty |= isPageDirty;
+        }
```
Comment (Nirus2000) [inline name.abuchen.portfolio.ui/src/name/abuchen/portfolio/ui/wizards/datatransfer/ImportController.java:42]:
```suggestion
        var isDirty = false;
        for (ReviewExtractedItemsPage page : pages)
        {
            page.afterPage();
            var isPageDirty = importPage(page, newSecurities);
            isDirty |= isPageDirty;
        }
```

**Review by Nirus2000 — CHANGES_REQUESTED** [review]

**Comment by Nirus2000** [comment]
This is very nice... i :heart_on_fire: it...

File: name.abuchen.portfolio.ui/src/name/abuchen/portfolio/ui/wizards/datatransfer/ExtractedEntry.java [hunk 2]
```diff
@@ -31,18 +45,23 @@ public void setImported(boolean isImported)
...
+        // while there is no security override
+        if (securityDependency != null && !securityDependency.isImported() && securityOverride == null)
+            return false;
+
         // otherwise import if either the status is OK or the user explicitly
         // overrides warnings
         return maxCode == Status.Code.OK
                         || (maxCode == Status.Code.WARNING && isImported != null && isImported.booleanValue()
-                        || (maxCode == Status.Code.WARNING && item.isInvestmentPlanItem()));
+                                        || (maxCode == Status.Code.WARNING && item.isInvestmentPlanItem()));
```
[the remark this answers is not in the document]
Comment (buchen) [reply name.abuchen.portfolio.ui/src/name/abuchen/portfolio/ui/wizards/datatransfer/ExtractedEntry.java:64]: It was only a formatting change - maybe I should remove even the formatting from the commit.

I am also not sure if the statements are identical due to the missing parenthesis.

**Comment by siprbaum** [comment]
On a technical level, I understand what the feature can do, but I am struggling a little for which use cases I would use it.

Would this be helpful e.g. when a security was merged/replaced by another?
E.g. in the past, the Lyxor MSCI World (WKN: ETF110) was merged/replaced by Amundi MSCI World (WKN: ETF018).
The bank automatically exchanged the existing ETF110 in the depot with Amundi ETF018 shares.

Any pointer to existing uses cases/documentation when to use this feature would be appreciated.

**Pull request merged**